File: examples/78-enum-api/main.py
"""Demonstrate enum constants available in the gopptx Python API.

This example demonstrates:
- ShapeType constants (RECTANGLE, ELLIPSE, ROUNDED_RECTANGLE, TRIANGLE)
- ConnectorType constants (STRAIGHT, ELBOW, CURVED)
- SlideLayoutType constants (BLANK, TITLE_ONLY, TITLE_CONTENT)
- ChartType constants (BAR, COLUMN, LINE, PIE)
- Shapes gallery using each ShapeType constant
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.constants import ConnectorType, ShapeType
from gopptx.presentation.charts import ChartType
from gopptx.presentation.slides import SlideLayoutType
from gopptx.schemas import Inches

_SHAPE_COLORS = ["4472C4", "C0504D", "9BBB59", "F79646", "8064A2", "4BACC6"]


def _add_core_shape_types_slide(prs: Presentation) -> None:
    """Gallery of core shape types."""
    idx = prs.add_slide("Core Shape Type Constants", layout=SlideLayoutType.TITLE_ONLY)
    types = [
        (ShapeType.RECTANGLE, "Rect"),
        (ShapeType.ROUNDED_RECTANGLE, "RoundRect"),
        (ShapeType.ELLIPSE, "Ellipse"),
        (ShapeType.TRIANGLE, "Triangle"),
    ]
    for i, (shape_type, label) in enumerate(types):
        col = i % 2
        row = i // 2
        prs.add_shape(
            idx,
            shape_type,
            (
                Inches(col * 4.5 + 0.5),
                Inches(row * 2.0 + 1.5),
                Inches(3.5),
                Inches(1.5),
            ),
            text=label,
            properties={"fill_color": _SHAPE_COLORS[i]},
        )


def _add_connector_types_slide(prs: Presentation) -> None:
    """Enumerate connector type constants."""
    idx = prs.add_slide("Connector Type Constants", layout=SlideLayoutType.TITLE_ONLY)
    connectors = [
        (ConnectorType.STRAIGHT, Inches(0.5), Inches(2.0), Inches(3.5), Inches(2.0)),
        (ConnectorType.ELBOW, Inches(0.5), Inches(3.0), Inches(4.5), Inches(4.5)),
        (ConnectorType.CURVED, Inches(5.0), Inches(2.0), Inches(9.0), Inches(4.5)),
    ]
    labels = ["STRAIGHT", "ELBOW", "CURVED"]
    for _label, (ctype, x1, y1, x2, y2) in zip(labels, connectors):
        prs.add_connector(idx, ctype, x1, y1, x2, y2)
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(1.2),
        Inches(9),
        Inches(0.5),
        text="ConnectorType.STRAIGHT / ELBOW / CURVED",
    )


def _add_layout_constants_slide(prs: Presentation) -> None:
    """Enumerate slide layout type constants."""
    prs.add_bullet_slide(
        "SlideLayoutType Constants",
        [
            f"BLANK          = {SlideLayoutType.BLANK!r}",
            f"TITLE_ONLY     = {SlideLayoutType.TITLE_ONLY!r}",
            f"TITLE_CONTENT  = {SlideLayoutType.TITLE_CONTENT!r}",
            "Use layout= parameter in prs.add_slide(title, layout=...)",
        ],
    )


def _add_chart_type_constants_slide(prs: Presentation) -> None:
    """Enumerate ChartType constants."""
    prs.add_bullet_slide(
        "ChartType Constants",
        [
            f"ChartType.BAR    = {ChartType.BAR!r}",
            f"ChartType.COLUMN = {ChartType.COLUMN!r}",
            f"ChartType.LINE   = {ChartType.LINE!r}",
            f"ChartType.PIE    = {ChartType.PIE!r}",
            "Pass to prs.add_chart(slide_idx, ChartType.X, ...)",
        ],
    )


def _add_shape_type_constants_slide(prs: Presentation) -> None:
    """List ShapeType and ConnectorType string values."""
    prs.add_bullet_slide(
        "ShapeType & ConnectorType Values",
        [
            f"ShapeType.RECTANGLE         = {ShapeType.RECTANGLE!r}",
            f"ShapeType.ELLIPSE           = {ShapeType.ELLIPSE!r}",
            f"ShapeType.ROUNDED_RECTANGLE = {ShapeType.ROUNDED_RECTANGLE!r}",
            f"ShapeType.TRIANGLE          = {ShapeType.TRIANGLE!r}",
            f"ConnectorType.STRAIGHT      = {ConnectorType.STRAIGHT!r}",
            f"ConnectorType.ELBOW         = {ConnectorType.ELBOW!r}",
            f"ConnectorType.CURVED        = {ConnectorType.CURVED!r}",
        ],
    )


def main() -> None:
    """Create presentation demonstrating enum constants."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    output_path = output_dir / "78_enum_api.pptx"

    with Presentation.new("Enum API Demo") as prs:
        _add_core_shape_types_slide(prs)
        _add_connector_types_slide(prs)
        _add_layout_constants_slide(prs)
        _add_chart_type_constants_slide(prs)
        _add_shape_type_constants_slide(prs)

        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print(
        "Demonstrated: ShapeType, ConnectorType, SlideLayoutType, ChartType constants"
    )
    print("  with a shapes gallery and enumeration slides")


if __name__ == "__main__":
    main()

File: examples/78-enum-api/main.go
// examples/78-enum-api demonstrates the enum constants available in gopptx.
//
// Covers shape type constants (geometry, arrows, callouts, flowchart, action
// buttons, math shapes), slide layout constants, placeholder type constants,
// chart type constructors (via the chart_compat layer), animation effect
// constants, and connector type constants.
//
// Run with: go run ./examples/78-enum-api/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/styling"
	log "github.com/djinn-soul/gopptx/pkg/stdlog"
)

const (
	outputDir  = "examples/output"
	outputFile = "78_enum_api.pptx"

	colorBlue = "4472C4"
	colorRed  = "C0504D"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	slides := []pptx.SlideContent{
		buildCoreShapeTypesSlide(),
		buildArrowTypesSlide(),
		buildLayoutConstantsSlide(),
		buildConnectorConstantsSlide(),
		buildAnimationConstantsSlide(),
		buildPlaceholderConstantsSlide(),
		buildTextFrameConstantsSlide(),
	}

	outputPath := filepath.Join(outputDir, outputFile)
	data, err := pptx.CreateWithSlides("Enum API Demo", slides)
	if err != nil {
		return fmt.Errorf("create presentation: %w", err)
	}
	if err := os.WriteFile(outputPath, data, 0o600); err != nil {
		return fmt.Errorf("write file: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

func buildCoreShapeTypesSlide() pptx.SlideContent {
	coreShapeTypes := []shapeConstant{
		{"Rect", pptx.ShapeTypeRectangle, colorBlue},
		{"RoundRect", pptx.ShapeTypeRoundedRectangle, colorRed},
		{"Ellipse", pptx.ShapeTypeEllipse, "9BBB59"},
		{"Triangle", pptx.ShapeTypeTriangle, "F79646"},
		{"Diamond", pptx.ShapeTypeDiamond, "8064A2"},
		{"Pentagon", pptx.ShapeTypePentagon, "4BACC6"},
		{"Hexagon", pptx.ShapeTypeHexagon, colorBlue},
		{"Parallelogram", pptx.ShapeTypeParallelogram, colorRed},
	}
	return buildShapeConstantsSlide("Core Shape Type Constants", coreShapeTypes)
}

func buildArrowTypesSlide() pptx.SlideContent {
	arrowTypes := []shapeConstant{
		{"RightArrow", pptx.ShapeTypeRightArrow, colorBlue},
		{"LeftArrow", pptx.ShapeTypeLeftArrow, colorRed},
		{"UpArrow", pptx.ShapeTypeUpArrow, "9BBB59"},
		{"DownArrow", pptx.ShapeTypeDownArrow, "F79646"},
		{"LR Arrow", pptx.ShapeTypeLeftRightArrow, "8064A2"},
		{"UD Arrow", pptx.ShapeTypeUpDownArrow, "4BACC6"},
		{"Quad", pptx.ShapeTypeQuadArrow, colorBlue},
		{"Bent", pptx.ShapeTypeBentArrow, colorRed},
	}
	return buildShapeConstantsSlide("Arrow Shape Constants", arrowTypes)
}

type shapeConstant struct {
	name      string
	shapeType string
	color     string
}

func buildShapeConstantsSlide(title string, shapeTypes []shapeConstant) pptx.SlideContent {
	slide := pptx.NewSlide(title)
	for i, shapeType := range shapeTypes {
		col := float64(i%4) * 2.3
		row := float64(i/4) * 1.8
		shape := pptx.NewShape(
			shapeType.shapeType,
			styling.Inches(col+0.3), styling.Inches(row+1.5),
			styling.Inches(2), styling.Inches(1.2),
		).
			WithFill(pptx.NewShapeFill(shapeType.color)).
			WithText(shapeType.name)
		slide = slide.AddShape(shape)
	}
	return slide
}

func buildLayoutConstantsSlide() pptx.SlideContent {
	return pptx.NewSlide("Slide Layout Constants").
		AddBullet(fmt.Sprintf(
			"SlideLayoutTitleAndContent    = %q", pptx.SlideLayoutTitleAndContent,
		)).
		AddBullet(fmt.Sprintf(
			"SlideLayoutTitleOnly          = %q", pptx.SlideLayoutTitleOnly,
		)).
		AddBullet(fmt.Sprintf(
			"SlideLayoutBlank              = %q", pptx.SlideLayoutBlank,
		)).
		AddBullet(fmt.Sprintf(
			"SlideLayoutCenteredTitle      = %q", pptx.SlideLayoutCenteredTitle,
		)).
		AddBullet(fmt.Sprintf(
			"SlideLayoutTitleAndBigContent = %q", pptx.SlideLayoutTitleAndBigContent,
		)).
		AddBullet(fmt.Sprintf(
			"SlideLayoutTwoColumn          = %q", pptx.SlideLayoutTwoColumn,
		))
}

func buildConnectorConstantsSlide() pptx.SlideContent {
	return pptx.NewSlide("Connector Type Constants").
		AddBullet(fmt.Sprintf("ConnectorTypeStraight = %q", pptx.ConnectorTypeStraight)).
		AddBullet(fmt.Sprintf("ConnectorTypeElbow    = %q", pptx.ConnectorTypeElbow)).
		AddBullet(fmt.Sprintf("ConnectorTypeCurved   = %q", pptx.ConnectorTypeCurved)).
		AddBullet(fmt.Sprintf("ArrowTypeNone     = %q", pptx.ArrowTypeNone)).
		AddBullet(fmt.Sprintf("ArrowTypeTriangle = %q", pptx.ArrowTypeTriangle)).
		AddBullet(fmt.Sprintf("ArrowTypeStealth  = %q", pptx.ArrowTypeStealth))
}

func buildAnimationConstantsSlide() pptx.SlideContent {
	return pptx.NewSlide("Animation Effect Constants").
		AddBullet(fmt.Sprintf("AnimationEntranceAppear  = %q", pptx.AnimationEntranceAppear)).
		AddBullet(fmt.Sprintf("AnimationEntranceFade    = %q", pptx.AnimationEntranceFade)).
		AddBullet(fmt.Sprintf("AnimationEntranceFlyIn   = %q", pptx.AnimationEntranceFlyIn)).
		AddBullet(fmt.Sprintf("AnimationEntranceZoom    = %q", pptx.AnimationEntranceZoom)).
		AddBullet(fmt.Sprintf("AnimationEntranceBounce  = %q", pptx.AnimationEntranceBounce)).
		AddBullet(fmt.Sprintf("AnimationEmphasisSpin    = %q", pptx.AnimationEmphasisSpin)).
		AddBullet(fmt.Sprintf("AnimationExitFadeOut     = %q", pptx.AnimationExitFadeOut)).
		AddBullet(fmt.Sprintf("AnimationOnClick         = %q", pptx.AnimationOnClick))
}

func buildPlaceholderConstantsSlide() pptx.SlideContent {
	return pptx.NewSlide("Placeholder Type Constants").
		AddBullet(fmt.Sprintf("PlaceholderTypeTitle    = %q", pptx.PlaceholderTypeTitle)).
		AddBullet(fmt.Sprintf("PlaceholderTypeBody     = %q", pptx.PlaceholderTypeBody)).
		AddBullet(fmt.Sprintf("PlaceholderTypeSubTitle = %q", pptx.PlaceholderTypeSubTitle)).
		AddBullet(fmt.Sprintf("PlaceholderTypeChart    = %q", pptx.PlaceholderTypeChart)).
		AddBullet(fmt.Sprintf("PlaceholderTypeTbl      = %q", pptx.PlaceholderTypeTbl)).
		AddBullet(fmt.Sprintf("PlaceholderTypePic      = %q", pptx.PlaceholderTypePic)).
		AddBullet(fmt.Sprintf("PlaceholderTypeMedia    = %q", pptx.PlaceholderTypeMedia))
}

func buildTextFrameConstantsSlide() pptx.SlideContent {
	return pptx.NewSlide("Text Frame Constants").
		AddBullet(fmt.Sprintf("TextAnchorTop    = %q", pptx.TextAnchorTop)).
		AddBullet(fmt.Sprintf("TextAnchorMiddle = %q", pptx.TextAnchorMiddle)).
		AddBullet(fmt.Sprintf("TextAnchorBottom = %q", pptx.TextAnchorBottom)).
		AddBullet(fmt.Sprintf("TextWrapNone   = %q", pptx.TextWrapNone)).
		AddBullet(fmt.Sprintf("TextWrapSquare = %q", pptx.TextWrapSquare)).
		AddBullet(fmt.Sprintf("TextAutoFitNone   = %q", pptx.TextAutoFitNone)).
		AddBullet(fmt.Sprintf("TextAutoFitShape  = %q", pptx.TextAutoFitShape)).
		AddBullet(fmt.Sprintf("TextAutoFitNormal = %q", pptx.TextAutoFitNormal))
}
